### PR TITLE
Product Security would like to query for binary RPM names for an SRPM name

### PR DESCRIPTION
**Pull request by blayson**
/rpms api endpoint for webapp, it allows to get rpm names by srpm_name_list and filter it by content_set_list, also it allows querying content_set by rpm_name_list

### CODE REVIEW

**Comment by matysek** [comment]
Renaming the endpoint to /package_names is also OK for us. The exact name does not matter much as long as it provides requested functionality.

File: webapp/rpm.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            repo_ids.extend(self.cache.repolabel2ids[label])
+        return repo_ids, repo_labels
+
+    def _packageids2names(self, pkg_id_list):
+        """For given list of pkg_ids returns list of package names"""
+        package_names = set()
+        for pkg_id in pkg_id_list:
+            pkg_name = self.cache.pkg_id2pkg_name[pkg_id]
+            package_names.add(pkg_name)
+        return list(package_names)
```
Comment (matysek) [inline webapp/rpm.py]: please use ```natsort.natsorted()``` for list of rpm names to have it properly aplhabetically sorted.

File: webapp/rpm.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                    pkg_ids = self.cache.pkg_name2pkg_ids[rpm]
+                    repo_ids = []
+                    for pid in pkg_ids:
+                        if pid in self.cache.pkgid2repoids:
+                            repo_ids.extend(self.cache.pkgid2repoids[pid])
+                    for rid in repo_ids:
+                        label = self.cache.repo_detail[rid][REPO_LABEL]
+                        content_data.setdefault(rpm, [])
+                        if label not in content_data[rpm]:
+                            content_data[rpm].append(label)
```
Comment (matysek) [inline webapp/rpm.py]: please use ```natsort.natsorted()``` for list of content_sets to have it properly aplhabetically sorted.

File: webapp/webapp.spec.yaml [hunk 3]
```diff
@@ -431,6 +431,26 @@ paths:
                 $ref: '#/components/schemas/VulnerabilitiesResponse'
       tags:
         - vulnerabilities
+
+  /v1/rpms:
+    post:
+      description: List of rpms by given srpm or rpm list and content set.
```
Comment (matysek) [inline webapp/webapp.spec.yaml:437]: I would prefer more self-explanatory description that mentions clearly it is about `names`.
e.g. ```List of content sets with associated rpm names by given srpm or rpm name list and content set.```

File: webapp/webapp.spec.yaml [hunk 4]
```diff
@@ -431,6 +431,26 @@ paths:
...
+      operationId: app.RpmsHandlerPost.post
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/RpmsRequest'
+      responses:
+        '200':
+          description: List of rpms or content sets by given srpm or rpm list and content set.
```
Comment (matysek) [inline webapp/webapp.spec.yaml:446]: I would prefer more self-explanatory description that mentions clearly it is about `names`.
e.g. ```List of content sets with associated rpm names by given srpm names or list content sets by given rpm names.```

File: webapp/webapp.spec.yaml [hunk 5]
```diff
@@ -1048,3 +1068,46 @@ components:
...
+        srpm_name_list:
+          type: array
+          items:
+            type: string
+            example: openssl
+        rpm_name_list:
+          type: array
+          items:
+            type: string
+            example: openssl
```
Comment (matysek) [inline webapp/webapp.spec.yaml:1084]: here as example rather use something like ```openssl-libs`` that it is clearly rpm name

File: webapp/webapp.spec.yaml [hunk 6]
```diff
@@ -1048,3 +1068,46 @@ components:
...
+            example: rhel-7-desktop-rpms
+
+    RpmsResponse:
+        properties:
+          last_change:
+            example: '2020-04-16 20:07:58.500192+00'
+          srpm_name_list:
+            type: object
+            properties:
+              openssl:
```
Comment (matysek) [inline webapp/webapp.spec.yaml:1098]: Is this typo or is it going to work to have ```openssl``` property? Probably the syntax should be similar to this?
```
srpm_name_list:
          type: object
          additionalProperties:
              type: object
              additionalProperties:
                  type: array
                  items:
                    type: string
                    example: 'openssl-libs'
```

File: webapp/webapp.spec.yaml [hunk 7]
```diff
@@ -431,6 +431,26 @@ paths:
...
+      responses:
+        '200':
+          description: List of rpms or content sets by given srpm or rpm list and content set.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/PackageNamesResponse'
+        <<: *format_resps
+      tags:
+        - rpms
```
Comment (matysek) [inline webapp/webapp.spec.yaml:453]: please also change this tag to 'package names' or similar.

**Comment by matysek** [comment]
I tested it and fails with non-existing content set. Tested request:
```yaml
{
  "srpm_name_list": [
    "openssl", "openssl-libs", "non-existing-srpm"
  ],
  "rpm_name_list": ["openssl", "openssl-libs", "non-existing-rpm"],
  "content_set_list": ["non-existing-content", "rhel-7-server-rpms"]
}
```
Failing output of webapp:
```
pp: [ERROR] Internal server error <8466827640>: please include this error id in bug report.'Traceback (most recent call last):\n  File "/git/app.py", line 75, in handle_request\n    res = api_endpoint.process_list(api_version, data)\n  File "/git/package_names.py", line 45, in process_list\n    content_set_labels.extend([label for label in content_set_list if\n  File "/git/package_names.py", line 46, in <listcomp>\n    self.cache.label2content_set_id[label] in content_set_ids])\nKeyError: \'non-existing-content\''|
64fc9816c0 2020-04-14 13:21:21,061 app: [ERROR] Input data for <877467640>: {'srpm_name_list': ['openssl', 'openssl-libs', 'non-existing-srpm'], 'rpm_name_list': ['openssl', 'openssl-libs', 'non-existing-rpm'], 'content_set_list': ['non-existing-content', 'rhel-7-server-rpms']}
```

**Comment by matysek** [comment]
It also does not behave correctly when a binary rpm is used in ```srpm_name_list```. It should behave like non-existing rpm.

request:
```yaml
{
"srpm_name_list": [
    "openssl", "openssl-devel", "binary-rpm", "non-existing-rpm"
  ]
}
```
Expected response:
```yaml
{
  "last_change": "2020-04-09T10:20:54.354523+00:00",
  "srpm_name_list": {
    "openssl": {
      "rhel-7-workstation-rpms": [
        "openssl",
        "openssl-devel",
        "openssl-libs"
      ],
      "rhel-7-server-rpms": [
        "openssl",
        "openssl-devel",
        "openssl-libs"
      ]
    },
  }
}
```

**Comment by matysek** [comment]
Also it should behave like this when an existing srpm is not in specified content_set:

request:
```yaml
{
"srpm_name_list": [
    "openssl", "existing-source-rpm-not-in-specified-content-set"
  ],
"content_set_list": [
   "specified-content-set"
]
}
```
expected response:
```yaml
{
  "last_change": "2020-04-09T10:20:54.354523+00:00",
  "srpm_name_list": {
    "openssl": {
      "specified-content-set": [
        "openssl",
        "openssl-devel",
        "openssl-libs"
      ],
    },
    "existing-source-rpm-not-in-specified-content-set": {}
  }
}
```

**Comment by jdobes** [comment]
rebased on master to fix travis ci test

**Comment by blayson** [comment]
I can split it into two different endpoints, it shouldn't take much time

**Comment by matysek** [comment]
I tested it and 2 of 3 things got fixed and 1 new identified.

* not fixed - for binary RPM in srpm_name_list it still returns binary rpm name in response - it should not.
* For this request
```yaml
{
"srpm_name_list": [
    "openssl", "samba", "samba-libs"
  ],
"content_set_list": [
   "rhel-7-workstation-rpms"
]
}
```
I get
```yaml
{
  "last_change": "2020-04-15T20:21:31.052083+00:00",
  "srpm_name_list": {
    "openssl": {
      "rhel-7-workstation-rpms": [
        "openssl",
        "openssl-devel",
        "openssl-libs"
      ]
    },
    "samba": {
      "rhel-7-workstation-rpms": [
        "libsmbclient",
        "libwbclient",
        "samba",
        "samba-client",
        "samba-client-libs",
        "samba-common",
        "samba-common-libs",
        "samba-common-tools",
        "samba-krb5-printing",
        "samba-libs",
        "samba-python",
        "samba-winbind",
        "samba-winbind-clients",
        "samba-winbind-modules"
      ]
    },
    "samba-libs": {
      "rhel-7-workstation-rpms": []
    }
  }
}
```
The package ```    "samba-libs": { "rhel-7-workstation-rpms": [] }``` should not be there because it is a binary rpm.

Another issue is what @jdobes already mentioned - api logic should for ```rpm_name_list``` take into account ```content_set_list```.
* For this request
```yaml
{
"rpm_name_list": [
    "samba-libs"
  ],
"content_set_list": [
   "rhel-9-blabla"
]
}
```
* I would expect this response - similar behaviour as for srpm_name_list
```yaml
{
  "last_change": "2020-04-15T20:21:31.052083+00:00",
  "rpm_name_list": {
    "samba-libs": [ ]
  }
}
```

**Comment by matysek** [comment]
Let's do this then:

1. fix srpm_name_list to ignore binary rpm names
2. split the endpoint into two - package_names/srpms and package_names/rpms as @jdobes  suggests. We are getting into too much complexity and confusion and as it turns out having it separate might be better option.
3. with this pull request implement only package_names/srpms
4. for package_names/rpms create another pull request
5. fix the behaviour for package_names/rpms to be similar to srpms api - respect ```content_set_list``` etc.

**Comment by blayson** [comment]
Is it necessarily to split it up into two different PRs? 
@matysek @jdobes

**Comment by matysek** [comment]
@blayson from my it is not necessary. However, consider the mood of stakeholders. You have a working deliverable that will be in master faster and stakeholders can start using it faster. You are happy something is done, stakeholders are happy it's ready.

**Comment by jdobes** [comment]
@blayson no need to split to 2 PRs, just depends how long will take to create both APIs, let's deliver at least srpms as soon as possible

**Comment by blayson** [comment]
All fixes done, currently the /package_names/srpms api should work as expected also I have add GET /package_names/srpms/{srpm}, it works like POST without any content_set_list specified

@jdobes I think it does not take a lot of time to create separate /rpms api endpoint, it almost done.  The question is do I need to do it here or create a separate PR?

**Pull request merged**